pep-3157.txt

```diff
@@ -0,0 +1,308 @@
+
+#. Another issue that has feels unintuitive to most is the lack of support for backslashes
+   within the expression component of an f-string. One example that keeps coming up is including
+   newline character in the expression part for joining containers. For example::
+
+    >>> a = ["hello", "world"]
+    >>> f"{'\n'.join(a)}"
+    File "<stdin>", line 1
+        f"{'\n'.join(a)}"
+                        ^
+    SyntaxError: f-string expression part cannot include a backslash
+
+   A common work-around for this was to eiter assign the newline to an intermediate variable or
+   pre-create the whole string prior to creating the f-string::
+
+    >>> a = ["hello", "world"]
+    >>> joined = '\n'.join(a)
+    >>> f"{joined}"
+    'hello\nworld'
+
+   It only feels natural to allow backslashes in the expression part now that the new PEG parser
+   can easily support it.
+
+    >>> a = ["hello", "world"]
+    >>> f"{'\n'.join(a)}"
+    'hello\nworld'
+
+
+#. What other weird choice did we made?
+
+
+TO BE DONE:
+
+* Describe why we are allowing crazy parenthesis nesting.
+* Describing how we are limiting nesting.
+* Describing other weird choices we made.
+
+Specification
+=============
+
+The formal proposed PEG grammar specification for f-strings is (see :pep:`617`
+for details on the syntax)::
+
+    fstring
+        | FSTRING_START fstring_middle* FSTRING_END
+    fstring_middle
+        | fstring_replacement_field
+        | FSTRING_MIDDLE
+    fstring_replacement_field
+        | '{' (yield_expr | star_expressions) "="? [ "!" NAME ] [ ':' fstring_format_spec* ] '}'
+    fstring_format_spec:
+        | FSTRING_MIDDLE
+        | fstring_replacement_field
+
+Three new tokens are introduced:
+
+* ``FSTRING_START``: This token includes f-string character (``f``/``F``) and the open quote(s).
+* ``FSTRING_MIDDLE``: This token includes everything between two expression braces (``}`` and ``{``).
+* ``FSTRING_END``: This token includes everything after the last expression brace (or the whole literal part
+if no expression exists) until the closing quote.
+
+These tokens are always string parts and they are semantically equivalent to the
+STRING token with the restrictions specified. These tokens must be produced by the lexer
+when lexing f-strings.  This means that **the tokenizer cannot produce a single token for f-strings anymore**. How
+the lexer emits this token is **not specified** as this will heavily depend on every
+implementation (even the Python version of the lexer in the standard library is
+implemented differently to the one used by the PEG parser).
+
+As an example, ``f'some words {a+b} more words {c+d} final words'`` will be tokenized as::
+
+    FSTRING_START - "f'"
+    FSTRING_MIDDLE - 'some words '
+    LBRACE - '{'
+    NAME - 'a'
+    PLUS - '+'
+    NAME - 'b'
+    RBRACE - '}'
+    FSTRING_MIDDLE
+    - , string='{'
+    NAME - 'c'
+    PLUS - '+'
+    NAME - 'd'
+    RBRACE - '}'
+    FSTRING_END - ' final words'
```

Comment:
Is there a typo here?

```suggestion
    FSTRING_START - "f'"
    FSTRING_MIDDLE - 'some words '
    LBRACE - '{'
    NAME - 'a'
    PLUS - '+'
    NAME - 'b'
    RBRACE - '}'
    FSTRING_MIDDLE - ' more words '
    LBRACE - '{'
    NAME - 'c'
    PLUS - '+'
    NAME - 'd'
    RBRACE - '}'
    FSTRING_END - ' final words' (without the end quote)
```

Also, regarding the end quotes not being part of the FSTRING_END token, we probably need to explain this, but also, is it really needed? I can't remember why that is the way it is.

Reply:
I don't think there is a reason, this is something we carried out from the draft. We probably should change it

Reply:
> Is there a typo here?

Yep, apologies there was a bad copy paste. Fixing it